### PR TITLE
XA Feature M1

### PR DESCRIPTION
**Pull request by a5anka**
This PR contains the implementation of WSO2 Message Broker to act as a resource manager according to the XA Transactions specification in accordance with [Java Transaction API (JTA) Specification 1.1](http://download.oracle.com/otndocs/jcp/jta-1.1-spec-oth-JSpec/?submit=Download). 

The AMQP version used for this is 0-9-1 and the implementation was done as a protocol extension. The message flow is similar to how it is done in AMQP version 0-10.

message-flow
Image extracted from AMQP 0-10 specification doc

### CODE REVIEW

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/DtxStore.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+package org.wso2.andes.kernel;
+
+import org.wso2.andes.store.HealthAwareStore;
+
+import java.util.List;
+import javax.transaction.xa.Xid;
+
+public interface DtxStore extends HealthAwareStore {
```
Comment (pamod) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/DtxStore.java]: Good if we could describe the role of DtxStore

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/MessagingEngine.java [hunk 2]
```diff
@@ -254,15 +259,34 @@ public void deleteMessages(List<DeliverableAndesMetadata> messagesToRemove) thro
             messagesOfStorageQueue.add(message);
             storageSeparatedMessages.put(message.getStorageQueueName(), messagesOfStorageQueue);
         }
+        return storageSeparatedMessages;
+    }
+
+    public void dtxCommit(long internalXid, List<AndesMessage> enqueueList, List<AndesAckData> ackDataList)
```
Comment (pamod) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/MessagingEngine.java]: Shall we add JavaDocs ?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/AckEventBatchHandler.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.andes.kernel.AndesAckData;
+import org.wso2.andes.kernel.disruptor.InboundEventHandler;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class AckEventBatchHandler extends InboundEventHandler {
```
Comment (pamod) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/AckEventBatchHandler.java]: Shall we add a JavaDoc explaining the role of the handler ?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/dtx/DtxBranch.java [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+
+    /**
+     * Current branch state
+     */
+    private State state = State.ACTIVE;
+
+    /**
+     * Used to keep the internal xid value used in message store
+     */
+    private long internalXid = NULL_XID;
```
Comment (pamod) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/dtx/DtxBranch.java]: Shall we have the instance variables defined at the beginning of the class.. altogether. Would be more readable

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/dtx/DtxBranch.java [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+    /**
+     * Getter for internalXid
+     */
+    public long getInternalXid() {
+        return internalXid;
+    }
+
+    public enum State {
+        SUSPENDED, ACTIVE, ROLLBACK_ONLY, PREPARED, FORGOTTEN, TIMED_OUT, HEUR_COM, HEUR_RB
```
Comment (pamod) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/dtx/DtxBranch.java]: Good if we could add comments mentioning the role of each state and it's applicability

File: modules/andes-core/broker/src/main/java/org/wso2/andes/amqp/QpidAndesBridge.java [hunk 6]
```diff
@@ -168,46 +168,14 @@ private QpidAndesBridge() {
...
      *
      * @param incomingMessage  message coming in
-     * @param channelID        id of the channel message came in
      * @param andesChannel     AndesChannel
      * @param transactionEvent not null if this is a message in a transaction, null otherwise
-     * @throws AMQException
+     * @throws AMQException when routing key is null
      */
-    public static void messageReceived(IncomingMessage incomingMessage, UUID channelID,
-                           AndesChannel andesChannel, InboundTransactionEvent transactionEvent) throws AMQException {
```
Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/amqp/QpidAndesBridge.java]: AndesChannel is removed from the signature of the method. Does this mean distributed transaction impl does  not have flow controlling?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesKernelBoot.java [hunk 7]
```diff
@@ -264,20 +265,27 @@ private static AndesContextStore createAndesContextStoreFromConfig() throws Exce
      * @return an implementation of {@link MessageStore}
      * @throws Exception if an error occurs
      */
-    private static MessageStore createMessageStoreFromConfig(AndesContextStore andesContextStore) throws Exception {
-        StoreConfiguration andesConfiguration = AndesContext.getInstance()
-                .getStoreConfiguration();
+    private static MessageStore createMessageStoreFromConfig(AndesContextStore andesContextStore,
+                                                             FailureObservingStoreManager failureObservingStoreManager)
```
Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesKernelBoot.java]: FailureObservingStoreManager - why do we need to pass it here?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/MessagingEngine.java [hunk 8]
```diff
@@ -203,24 +209,7 @@ public void moveMessageToDeadLetterChannel(DeliverableAndesMetadata messageToRem
...
-            storageSeparatedMessages.put(message.getStorageQueueName(), messagesOfStorageQueue);
-        }
-
-        //delete message content along with metadata
-        for (Map.Entry<String, List<AndesMessageMetadata>> entry : storageSeparatedMessages.entrySet()) {
-            messageStore.deleteMessages(entry.getKey(), entry.getValue());
-        }
-
-        //TODO:message can be in delivery path. If so we need to decrement slot message count
+        messageStore.deleteMessages(messagesToRemove);
```
Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/MessagingEngine.java]: This is good. We can get rid of unnecessary for loop passing the container.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/AckEventBatchHandler.java [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+    private static Log log = LogFactory.getLog(AckHandler.class);
+    private final List<AndesAckData> ackDataList;
+    private final int turn;
+    private final int groupCount;
+    private final int batchSize;
+    private final AckHandler ackHandler;
+
+    public AckEventBatchHandler(int turn, int groupCount, int batchSize, AckHandler ackHandler) {
```
Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/AckEventBatchHandler.java]: I believe we batched ack earlier as well. What is the change in new design?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/InboundEventContainer.java [hunk 10]
```diff
@@ -194,6 +213,9 @@ public void updateState() throws AndesException {
             case TRANSACTION_CLOSE_EVENT:
                 getTransactionEvent().updateState();
                 break;
+            case DTX_COMMIT_EVENT:
+                getDtxBranch().updateState();
+                break;
             case PUBLISHER_RECOVERY_EVENT:
                 executeRecoveryEvent();
                 break;
```
Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/InboundEventContainer.java]: What about transaction start and other commands? Do not they go through this?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/InboundEventManager.java [hunk 11]
```diff
@@ -160,14 +166,14 @@ public InboundEventManager(AndesSubscriptionManager subscriptionManager,
...
-        disruptor.after(preProcessor).handleEventsWith(concurrentBatchEventHandlers);
-
-        // State event handler update the state of Andes after other handlers work is done.
-        // State event handler will execute last. This handler will clear the event container.
-        disruptor.after(concurrentBatchEventHandlers).handleEventsWith(stateEventHandler);
+        disruptor.after(preProcessor)
+                .handleEventsWith(batchEventHandlers)
+                .then(stateEventHandler);   // State event handler update the state of Andes after other handlers work
+                                            // is done. State event handler will execute last. This handler will clear
+                                            // the event container.
```
Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/InboundEventManager.java]: It is better to add an inline doc about the order of the handlers

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/dtx/AlreadyKnownDtxException.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+ */
+
+package org.wso2.andes.kernel.dtx;
+
+import org.wso2.andes.server.txn.DtxException;
+
+import javax.transaction.xa.Xid;
+
+
+public class AlreadyKnownDtxException extends DtxException
```
Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/dtx/AlreadyKnownDtxException.java]: AlreadyKnown? You mean defined exceptions?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/dtx/DistributedTransaction.java [hunk 13]
```diff
@@ -0,0 +1,208 @@
...
+    }
+
+    /**
+     * Mark transaction as failed. We will fail the transaction at prepare stage if this was called
+     *
+     * @param reason reason for failure
+     */
+    public void failTransaction(String reason) {
+        // We don't need to keep track of errors happening the transaction is not active
+        if (isInsideStartEnd()) {
```
Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/dtx/DistributedTransaction.java]: isInTxBountry()?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/dtx/DtxBranch.java [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+import org.wso2.andes.kernel.slot.SlotMessageCounter;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.transaction.xa.Xid;
+
+public class DtxBranch implements AndesInboundStateEvent {
```
Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/dtx/DtxBranch.java]: Better to specify what is represented by a DtxBranch

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/dtx/DtxRegistry.java [hunk 15]
```diff
@@ -0,0 +1,246 @@
...
+    public synchronized boolean registerBranch(DtxBranch branch) {
+        ComparableXid xid = new ComparableXid(branch.getXid());
+        if(!branches.containsKey(xid)) {
+            branches.put(xid, branch);
+            return true;
+        }
+        return false;
+    }
+
+    public synchronized void prepare(Xid xid)
```
Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/dtx/DtxRegistry.java]: Can many threads all prepare for a given xid?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/server/handler/DtxEndHandler.java [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+    public static DtxEndHandler getInstance() {
+        return _instance;
+    }
+
+    private DtxEndHandler() {
+    }
+
+    @Override
+    public void methodReceived(AMQStateManager stateManager, DtxEndBodyImpl body, int channelId) throws AMQException {
+        Xid xid = new XidImpl(body.getBranchId(), body.getFormat(), body.getGlobalId());
```
Comment (abeykoon) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/server/handler/DtxEndHandler.java]: Is this going by Spec?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/MessagingEngine.java [hunk 17]
```diff
@@ -280,16 +304,8 @@ public void deleteMessagesById(List<Long> messagesToRemove) throws AndesExceptio
...
-        for (DeliverableAndesMetadata message : messagesToMove) {
-            List<AndesMessageMetadata> messagesOfStorageQueue = storageSeparatedMessages
-                    .get(message.getStorageQueueName());
-            if (null == messagesOfStorageQueue) {
-                messagesOfStorageQueue = new ArrayList<>();
-            }
-            messagesOfStorageQueue.add(message);
-            storageSeparatedMessages.put(message.getStorageQueueName(), messagesOfStorageQueue);
-        }
+        Map<String, List<AndesMessageMetadata>> storageSeparatedMessages = groupByStorageQueue(
```
Comment (sdkottegoda) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/MessagingEngine.java]: I think it's not needed to group by storage queues. IIRC, we move the messages to DLC using the message id anyways.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/ConcurrentBatchEventHandler.java [hunk 18]
```diff
@@ -104,14 +113,16 @@ public ConcurrentBatchEventHandler(long turn, int groupCount, int batchSize,
...
-        
-        
+
         // Batch only relevant event type. Skip the rest
         if (eventType == event.getEventType()) {
 
             currentTurn = sequence % groupCount;
             if (turn == currentTurn) {
-                eventList.add(event);
+                messageList.addAll(event.getMessageList());
```
Comment (sdkottegoda) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/ConcurrentBatchEventHandler.java]: Can an event have a list of messages?

Comment (Asitha) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/ConcurrentBatchEventHandler.java]: Yes, In topic case where we have clones.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/AckHandler.java [hunk 19]
```diff
@@ -74,21 +73,20 @@
         FailureObservingStoreManager.registerStoreHealthListener(this);
     }
 
-    @Override
-    public void onEvent(final List<InboundEventContainer> eventList) throws Exception {
+    public void onEvent(final List<AndesAckData> ackDataList) throws Exception {
```
Comment (sdkottegoda) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/AckHandler.java]: Just my thought, wouldn't it be less confusing if we rename the method to something like processAckList or something. The method name makes me think it's a disruptor handler. Same goes with other similar classes

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/InboundEventContainer.java (hunk 10 above)
Comment (a5anka) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/InboundEventContainer.java]: No. dtx.start and dtx.end are handled at the AMQP transport level. No need to expose them to andes-core.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/dtx/AlreadyKnownDtxException.java (hunk 12 above)
Comment (a5anka) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/dtx/AlreadyKnownDtxException.java]: :smile: This is thrown when the we try to start a already started transaction. Can change the name if this is confusing. WDYT?

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/AckEventBatchHandler.java (hunk 9 above)
Comment (Asitha) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/AckEventBatchHandler.java]: We batched the inbound events earlier through the concurrent batch handler and from that we batched the acknowledgments through Ackhandler and processed it. Now we are directly batching the acknowledgments.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/dtx/DtxRegistry.java (hunk 15 above)
Comment (a5anka) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/dtx/DtxRegistry.java]: Generally, multiple threads won't call prepare simultaneously. But we need to make sure the branch states are modified atomically.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/server/handler/DtxEndHandler.java (hunk 16 above)
Comment (a5anka) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/server/handler/DtxEndHandler.java]: Yes

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesKernelBoot.java (hunk 7 above)
Comment (Asitha) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/AndesKernelBoot.java]: FailureObservingStoreManager is used by the dtxStore which is created by the MessageStore. Therefore we need to pass it.

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/InboundEventManager.java (hunk 11 above)
Comment (Asitha) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/disruptor/inbound/InboundEventManager.java]: There is already a comment above with the order of execution of handlers

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/dtx/DtxBranch.java (hunk 4 above)
Comment (a5anka) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/dtx/DtxBranch.java]: Fixed in https://github.com/wso2/andes/pull/768/commits/c7b9fdb21ea17a2f3a95ee41c7bcef2308881efb

File: modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/MessagingEngine.java (hunk 17 above)
Comment (Asitha) [inline modules/andes-core/broker/src/main/java/org/wso2/andes/kernel/MessagingEngine.java]: Ok. We will update this in the next iteration.

**Pull request merged**